contracts/Core/StakeManager.sol

```diff
@@ -173,44 +176,65 @@ contract StakeManager is Initializable, StakeStorage, StateManager, Pause, Stake
+        require(locks[msg.sender][stakers[stakerId].tokenAddress][LockType.Unstake].amount == 0, "Existing Unstake Lock");
+
+        uint32 epoch = _getEpoch(epochLength);
+        Structs.Staker storage staker = stakers[stakerId];
+        IStakedToken sToken = IStakedToken(staker.tokenAddress);
+
+        require(sToken.balanceOf(msg.sender) >= sAmount, "Invalid Amount");
+
+        locks[msg.sender][staker.tokenAddress][LockType.Unstake] = Structs.Lock(sAmount, 0, epoch + unstakeLockPeriod);
+        emit Unstaked(msg.sender, epoch, stakerId, sAmount, staker.stake, block.timestamp);
+
+        require(sToken.transferFrom(msg.sender, address(this), sAmount), "sToken transfer failed");
+    }
+
+    function initiateWithdraw(uint32 stakerId) external initialized whenNotPaused {
         State currentState = _getState(epochLength);
         require(currentState != State.Propose, "Unstake: NA Propose");
         require(currentState != State.Dispute, "Unstake: NA Dispute");
+
         uint32 epoch = _getEpoch(epochLength);
+        require(stakerId != 0, "staker doesnt exist");
+        require(locks[msg.sender][stakers[stakerId].tokenAddress][LockType.Unstake].unlockAfter != 0, "Did not unstake");
+        require(locks[msg.sender][stakers[stakerId].tokenAddress][LockType.Unstake].unlockAfter <= epoch, "Withdraw epoch not reached");
+        require(
+            locks[msg.sender][stakers[stakerId].tokenAddress][LockType.Unstake].unlockAfter + withdrawInitiationPeriod >= epoch,
+            "Initiation Period Passed"
+        ); // Can Use ExtendLock
 
         Structs.Staker storage staker = stakers[stakerId];
-        require(staker.id != 0, "staker.id = 0");
-        require(staker.stake > 0, "Nonpositive stake");
-        require(locks[msg.sender][staker.tokenAddress].amount == 0, "Existing Lock");
-        require(sAmount > 0, "Non-Positive Amount");
+        IStakedToken sToken = IStakedToken(staker.tokenAddress);
+        Structs.Lock storage lock = locks[msg.sender][staker.tokenAddress][LockType.Unstake];
 
         // slither-disable-next-line reentrancy-events,reentrancy-no-eth
         rewardManager.giveInactivityPenalties(epoch, stakerId);
 
-        IStakedToken sToken = IStakedToken(staker.tokenAddress);
-        require(sToken.balanceOf(msg.sender) >= sAmount, "Invalid Amount");
-
-        uint256 rAmount = _convertSRZRToRZR(sAmount, staker.stake, sToken.totalSupply());
+        uint256 rAmount = _convertSRZRToRZR(lock.amount, staker.stake, sToken.totalSupply());
         staker.stake = staker.stake - rAmount;
 
         // Transfer commission in case of delegators
         // Check commission rate >0
         uint256 commission = 0;
         if (stakerIds[msg.sender] != stakerId && staker.commission > 0) {
             // Calculate Gain
-            uint256 initial = sToken.getRZRDeposited(msg.sender, sAmount);
+            uint256 initial = sToken.getRZRDeposited(msg.sender, lock.amount);
```

Comment:
hmm this won't work as intended now 
as msg.sender don't own lock.amount now, becoz now we are transferring it in unstake to this contract.

For ex.
```
Razor depo 100
sRZR 100
Current sRZR valuation 200
```
now you are transferring it in unstake to this contract 
hence for this user 
if you unstake 100 sRZRs 
then razorDeposited for you is updated to 0 
and for this contract as 100 

https://github.com/razor-network/contracts/blob/master/contracts/tokenization/StakedToken.sol#L70

then when you calculate gain
its consider as 
current sRZR value(200) - razorDeposited(0) = 200
and commission is calculated on that 
not on 
200 - 100

Reply:
I am not sure why this was not picked up in tests itself
Commission calculation will always go wrong with this change,
Are we not calculating commission in tests ?

Reply:
We are calculating commission in tests. such an error never popped up

Reply:
resolved?